pkg/inst-api-semconv/instrumenter/http/http_attrs_extractor.go

```diff
@@ -0,0 +1,118 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package http
+
+import (
+	"context"
+	"github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/inst-api-semconv/instrumenter/utils"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.30.0"
+	"sync/atomic"
+)
+
+type HttpRequest interface{}
+
+type HttpResponse interface{}
+
+type HttpCommonAttrsExtractor[REQUEST HttpRequest, RESPONSE HttpResponse, GETTER1 HttpCommonAttrsGetter[REQUEST, RESPONSE]] struct {
+	HttpGetter       GETTER1
+	AttributesFilter func(attrs []attribute.KeyValue) []attribute.KeyValue
+}
+
+func (h *HttpCommonAttrsExtractor[REQUEST, RESPONSE, GETTER1]) OnStart(attributes []attribute.KeyValue, parentContext context.Context, request REQUEST) ([]attribute.KeyValue, context.Context) {
+	attributes = append(attributes, attribute.KeyValue{
+		Key:   semconv.HTTPRequestMethodKey,
+		Value: attribute.StringValue(h.HttpGetter.GetRequestMethod(request)),
+	})
+	return attributes, parentContext
+}
+
+func (h *HttpCommonAttrsExtractor[REQUEST, RESPONSE, GETTER]) OnEnd(attributes []attribute.KeyValue, context context.Context, request REQUEST, response RESPONSE, err error) ([]attribute.KeyValue, context.Context) {
+	statusCode := h.HttpGetter.GetHttpResponseStatusCode(request, response, err)
+	attributes = append(attributes, attribute.KeyValue{
+		Key:   semconv.HTTPResponseStatusCodeKey,
+		Value: attribute.IntValue(statusCode),
+	})
+	errorType := h.HttpGetter.GetErrorType(request, response, err)
+	if errorType != "" {
+		attributes = append(attributes, attribute.KeyValue{Key: semconv.ErrorTypeKey, Value: attribute.StringValue(errorType)})
+	}
+	return attributes, context
+}
+
+type HttpClientAttrsExtractor[REQUEST HttpRequest, RESPONSE HttpResponse, GETTER1 HttpClientAttrsGetter[REQUEST, RESPONSE]] struct {
+	Base HttpCommonAttrsExtractor[REQUEST, RESPONSE, GETTER1]
+}
+
+func (h *HttpClientAttrsExtractor[REQUEST, RESPONSE, GETTER1]) OnStart(attributes []attribute.KeyValue, parentContext context.Context, request REQUEST) ([]attribute.KeyValue, context.Context) {
```

Comment:
It is appreciated to document the behaviour of the function using comments for better readability.